File: server/services/identity-service/SafeRide.Identity/SafeRide.Identity.Api/Common/ApiResponse.cs
namespace SafeRide.Identity.Api.Common;

public class ApiResponse<T>
{
    public bool Success { get; init; }
    public T? Data { get; init; }
    public string? Message { get; init; }
    public ApiError? Error { get; init; }

    public static ApiResponse<T> Ok(T data, string? message = null) =>
        new()
        {
            Success = true,
            Data = data,
            Message = message,
        };

    public static ApiResponse<T> Fail(string code, string message) =>
        new() { Success = false, Error = new ApiError(code, message) };
}

public sealed record ApiError(string Code, string Message);

File: server/services/bus-service/bus/src/main/java/com/saferide/bus/dto/ApiResponse.java
package com.saferide.bus.dto;

public record ApiResponse<T>(boolean success, T data, String message, ApiError error) {

    public record ApiError(String code, String message) {}

    public static <T> ApiResponse<T> ok(T data, String message) {
        return new ApiResponse<>(true, data, message, null);
    }

    public static <T> ApiResponse<T> ok(T data) {
        return ok(data, null);
    }

    public static <T> ApiResponse<T> fail(String code, String message) {
        return new ApiResponse<>(false, null, null, new ApiError(code, message));
    }
}
